apps/charge-smart-wallets-service/src/data-layer/data-layer.service.ts

```diff
@@ -12,54 +12,58 @@ import {
 import { isNil } from 'lodash'
 import { TokenService } from '@app/smart-wallets-service/common/services/token.service'
 import { TokenTransferWebhookDto } from '@app/smart-wallets-service/smart-wallets/dto/token-transfer-webhook.dto'
+import { SmartWalletsEventsService } from '../smart-wallets/smart-wallets-events.service'
 
 @Injectable()
 export class DataLayerService {
   private readonly logger = new Logger(DataLayerService.name)
 
-  constructor (
+  constructor(
     @Inject(userOpString)
     private userOpModel: Model<UserOp>,
     @Inject(walletActionString)
     private paginatedWalletActionModel: PaginateModel<WalletActionDocument>,
     private userOpFactory: UserOpFactory,
-    private tokenService: TokenService
+    private tokenService: TokenService,
+    private smartWalletsEventsService: SmartWalletsEventsService
   ) { }
 
-  async recordUserOp (baseUserOp: BaseUserOp) {
+  async recordUserOp(baseUserOp: BaseUserOp) {
     const userOp = await this.userOpFactory.createUserOp(baseUserOp)
-
     const response = this.userOpModel.create(userOp)
     this.createWalletActionFromUserOp(userOp)
     return response
   }
 
-  async updateUserOp (body: UserOp) {
+  async updateUserOp(body: UserOp) {
     const existingUserOp = await this.userOpModel.findOne({ userOpHash: body.userOpHash })
     if (isNil(existingUserOp)) {
       return 'No record found with the provided userOpHash'
     }
     const updatedUserOp = await this.userOpModel.findOneAndUpdate({ userOpHash: body.userOpHash }, body, { new: true })
-
+    this.smartWalletsEventsService.publishUserOp(updatedUserOp.sender, updatedUserOp)
     this.updateWalletAction(updatedUserOp)
     return updatedUserOp
   }
 
-  async createWalletActionFromUserOp (parsedUserOp: UserOp) {
+  async createWalletActionFromUserOp(parsedUserOp: UserOp) {
     try {
       const walletAction = await parsedUserOpToWalletAction(parsedUserOp, this.tokenService)
+      this.smartWalletsEventsService.publishWalletAction(walletAction.walletAddress, walletAction)
       return this.paginatedWalletActionModel.create(walletAction)
     } catch (error) {
       console.log(error)
     }
   }
 
-  async updateWalletAction (userOp: any) {
+  async updateWalletAction(userOp: any) {
     const walletAction = confirmedUserOpToWalletAction(userOp)
-    return this.paginatedWalletActionModel.findOneAndUpdate({ userOpHash: walletAction.userOpHash }, walletAction)
+    const updatedWalletAction = await this.paginatedWalletActionModel.findOneAndUpdate({ userOpHash: walletAction.userOpHash }, walletAction, { new: true }).lean() as any
```

Comment:
why do you do `as any` here?
seems like a bad practice. Can you use the specific type?